clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2699,6 +2898,52 @@ namespace {
       if (Op == UnaryOperatorKind::UO_Deref)
         return CreateBoundsInferenceError();
 
+      // Update UEQ and G for inc/dec operators `++e1`, `e1++`, `--e1`, `e1--`.
+      // At this point, State contains UEQ and G for `e1`.
+      if (UnaryOperator::isIncrementDecrementOp(Op)) {
+        // Create the target of the implied assignment `e1 = e1 +/- 1`.
+        Expr *Target = CreateImplicitCast(SubExpr->getType(),
+                                            CK_LValueToRValue, SubExpr);
+
+        // Only use the RHS `e1 +/1 ` of the implied assignment to update
+        // UEQ and G if the integer constant 1 can be created, which is
+        // only true if `e1` has integer type or integer pointer type.
+        IntegerLiteral *One = CreateIntegerLiteral(1, SubExpr->getType());
+        Expr *RHS = nullptr;
+        if (One) {
+          BinaryOperatorKind RHSOp = UnaryOperator::isIncrementOp(Op) ?
+                                      BinaryOperatorKind::BO_Add :
+                                      BinaryOperatorKind::BO_Sub;
+          RHS = ExprCreatorUtil::CreateBinaryOperator(S, SubExpr, One, RHSOp);
+        }
+
+        // Update UEQ for inc/dec operators where `e1` is a variable.  Any
+        // expressions in UEQ that use the value of `e1` need to be adjusted
+        // using the original value of `e1`, since `e1` has been updated.
+        if (DeclRefExpr *V = GetLValueVariable(SubExpr)) {
+          // Clear G before updating UEQ since G currently contains
+          // expressions that produce the same value as the variable `e1`,
+          // and these expressions should not be added to UEQ.
+          State.G.clear();
+          Expr *OV = GetOriginalValue(V, RHS, State.UEQ);
+          UpdateAfterAssignment(V, Target, OV, CSS, State, State);
+        }
+
+        // Update the set G of expressions that produce the same value as e.
+        if (RHS) {
+          // For post-inc/dec operators, use the rvalue for `e1` to update G.
```

Comment:
This logic is reversed.   In the state after the expression ++x has executed, the non-modifying expression equivalent to the value produced by ++x is x.   For x++, the non-modifying expression is x - 1..